source/Conversation.cpp

```diff
@@ -442,46 +459,119 @@ const Sprite *Conversation::Scene(int node) const
 
 
 // Find out where the conversation goes if the given option is chosen.
-int Conversation::NextNode(int node, int choice) const
+int Conversation::NextNode(int node, int element) const
 {
-	if(static_cast<unsigned>(node) >= nodes.size()
-			|| static_cast<unsigned>(choice) >= nodes[node].data.size())
+	if(!NodeIsValid(node) || !ElementIsValid(node, element))
 		return DECLINE;
 
-	return nodes[node].data[choice].second;
+	return nodes[node].elements[element].next;
+}
+
+
+
+// Returns whether the given node should be displayed.
+bool Conversation::ShouldShowText(const map<string, int64_t> &vars, int node, int element) const
+{
+	if(!NodeIsValid(node))
+		return false;
+	else if(IsChoice(node) ? !ElementIsValid(node, element) : element != 0)
+		return false;
+	const auto &data = nodes[node].elements[element];
+	if(data.conditions.IsEmpty())
+		return true;
+	return data.conditions.Test(vars);
+}
+
+
+
+// Returns true if the given node index is in the range of valid nodes for this
+// Conversation.
+bool Conversation::NodeIsValid(int node) const
+{
+	if(node < 0)
+		return false;
+	return static_cast<unsigned>(node) < nodes.size();
+}
+
+
+
+// Returns true if the given node index is in the range of valid nodes for this
+// Conversation *and* the given element index is in the range of valid elements
+// for the given node.
+bool Conversation::ElementIsValid(int node, int element) const
+{
+	if(!NodeIsValid(node))
+		return false;
+	else if(element < 0)
+		return false;
+	return static_cast<unsigned>(element) < nodes[node].elements.size();
 }
 
 
 
-// Parse the children of the given node to see if then contain any "gotos."
-// If so, link them up properly. Return true if gotos were found.
+// Parse the children of the given node to see if then contain any "gotos," or
+// "to shows." If so, link them up properly. Return true if gotos or
+// conditions were found.
 bool Conversation::LoadGotos(const DataNode &node)
 {
 	bool hasGoto = false;
+	bool hasCondition = false;
 	for(const DataNode &child : node)
 	{
-		if(hasGoto)
+		if(hasGoto && hasCondition)
```

Comment:
Do we still need this check here if we already check for double goto's and double to-display's? The last check in the else-part would also seem to catch the issues that are otherwise caught here.

Reply:
No, the check shouldn't be there because it will mask the more meaningful `"Warning: Ignoring extra condition in conversation choice:"` error.

Reply:
I removed that "if" entirely, because it can hide the more specific error messages. I'm only 98% certain of that, so if people could skim over that code again, I'd appreciate it.

It'll break if you get rid of `&& hasCondition` and retain `hasGoto` though. You need either both or neither for the code to work.